doc/contributing/fpga/debugging_with_ila.md

```diff
@@ -0,0 +1,280 @@
+In that case, add a `parameter logic` to the module with a default value of `1'b0`, set its value to `1'b1` for the instance in which you would like to instantiate the ILA, and use that parameter to conditionally instantiate the ILA.
+For example, to debug `dmi_jtag` in `rv_dm` only, change `dmi_jtag.sv` like this:
+```sv
+module dmi_jtag #(
+  parameter logic [31:0] IdcodeValue = 32'h00000001,
+  parameter logic DebugIla = 1'b0 // add this
+) (
+  /* ... */
+
+  if (DebugIla) begin : gen_ila // wrap instance into this `if (...) begin end`
+    ila_0 u_ila_0 (
+      /* ... */
+    );
+  end
+);
+```
+And change the instantiation of `dmi_jtag` in `rv_dm` like this:
+```sv
+dmi_jtag #(
+  .IdcodeValue (IdcodeValue),
+  .DebugIla    (1'b1) // add this
+) dap (
+  /* ... */
+);
+```
+
+If you need to debug signals in different modules, you can either instantiate one ILA in each of them or route signals from other modules into one in which you have already instantiated an ILA.
+Routing is more effort than instantiating multiple ILAs, but it may be required to sample signals from different modules in the same clock cycle.
+If you need to debug signals from many different modules, you may want to instantiate one ILA at the lowest common design hierarchy and route the signals there.
+
+
+## Parametrizing the ILA and getting it synthesized in Vivado
+
+Next, we need to create an ILA *IP core* from Vivado's library, parametrize it, and synthesize it before synthesizing the rest of our design.
+To execute commands in Vivado before synthesis of the Earlgrey design, we have to add them to `hw/top_earlgrey/util/vivado_hook_synth_design_pre.tcl`.
+The commands to create and configure an ILA core can be obtained from Vivado's GUI, although we have to tweak them to work with our flow, which uses Vivado *in-memory projects*.
+
+Open Vivado's GUI (creating a project is not necessary), click on *Window* -> *IP Catalog*, and search for *ILA*.
+Double-click on *ILA (Integrated Logic Analyzer)* to open the customization window.
+Set *Component Name* to the name of the module in RTL (`ila_0` in our example above).
+In the *General Options* tab:
+- Keep *Monitor Type* to *Native*.
+- Keep *Number of Probes* to 1 unless you have added additional probe ports (`probe1` etc.) in RTL.
+- Select *Sample Data Depth* according to your needs; this determines how many samples can be saved per trigger action.
+  4096 should be feasible unless you are probing many signals.
+- Keep *Same Number of Comparators for All Probe Ports* ticked and *Number of Comparators* at 1 unless you know you need different settings.
+- If you can define a trigger condition based on probe signals, leave *Trigger In Port* unticked.
+  Otherwise, i.e., if you need to trigger the ILA from RTL, tick it and connect the added `trig_in` and `trig_in_ack` ports (see diagram on the left of the GUI window with *Show disabled ports* unticked).
```

Comment:
```suggestion
  If you need to trigger the ILA from RTL or otherwise, tick it and connect the added `trig_in` and `trig_in_ack` ports (see diagram on the left of the GUI window with *Show disabled ports* unticked).
```